File: pre-commit-tools/update_hooks.py
# -*- coding: utf-8 -*-
"""Updates all hooks and their `additional_dependencies`."""


import json
import requests
import re
import os
from config import (
    pre_commit_config_name,
    global_pre_commit_config_location,
)


print("Running autoupdate")
os.system("pre-commit autoupdate")

print("Running garbage collection")
os.system("pre-commit gc")


print("Updating `additional_dependencies` in .pre-commit-config.yaml")

with open(global_pre_commit_config_location) as file:
    data = file.read()

lines = data.split("\n")
result_lines = []
search_for_dependencies = False
extract_dependency_regex = re.compile(
    r"\s+- (?P<name>@?[\w-]+)@(?P<version>[\d.^~<=>]+)"
)
replace_version_regex = re.compile(r"(?P<base>\s+- .*@)(?:[\d.^~<=>]+)")


def get_request_url(dependency_name):
    """Format API request URL from dependency_name.

    Args:
        dependency_name:
            Dependency Name

    Returns:
        Formatted request URL
    """
    return f"https://registry.npmjs.org/{dependency_name}/"


def get_latest_version(dependency_name):
    """Fetch the latest version of a dependency

    Args:
        dependency_name:
            Dependency Name

    Returns:
        The version string
    """
    request = requests.get(get_request_url(dependency_name))
    if request.status_code != 200:
        raise Exception(
            f"Failed to fetch metaData for {dependency_name}"
        )
    return request.json()["dist-tags"]["latest"]


for line in lines:

    if line == "":
        search_for_dependencies = False

    if search_for_dependencies and "@" in line:
        match = extract_dependency_regex.search(line)
        if match:
            dependency_name = match.group("name")
            current_version = match.group("version")
            latest_version = get_latest_version(dependency_name)
            if f"^{latest_version}" != current_version:
                line = replace_version_regex.sub(
                    fr"\g<base>^{latest_version}", line
                )
                print(
                    f"Updating {dependency_name} from {current_version}"
                    f" to {latest_version}"
                )
            else:
                print(
                    f"{dependency_name}@{current_version} is already"
                    f" up to date"
                )

    if "additional_dependencies:" in line:
        search_for_dependencies = True

    result_lines.append(line)

final_data = "\n".join(result_lines)
with open(global_pre_commit_config_location, "w") as file:
    file.write(final_data)

File: pre-commit-tools/updateHooks.js
/**
 * Updates all hooks and their `additional_dependencies`
 */

import fs from 'fs';
import fetch from 'node-fetch';


const configFileLocation = process.argv[2];
const config = fs.readFileSync(configFileLocation).toString();

console.log(`Updating hooks and NPM \`additional_dependencies\` in ${configFileLocation}`);

const reExtractRepo = /- repo: https:\/\/github\.com\/(?<gitHubRepo>[^\n]+)/;
const reRev = /(?<base>rev: )(?<version>[^\n]+)/;
const reExtractDependency = /\s+- (?<dependencyName>@?[\w-]+)@(?<currentVersion>[\d.^~<=>]+)/;
const reReplaceVersion = /(?<base>\s+- .*@)(?:[\d.^~<=>]+)/;

const get = async (url)=>
  fetch(url, 'GITHUB_TOKEN' in process.env ? {
    headers: {
      Authorization: `token ${process.env.GITHUB_TOKEN}`,
    },
  } : {})
    .then(response=>new Promise((resolve, reject)=>
      response.json().then(data=>resolve([response,data]))
    ))
    .then(([{status}, data])=>{
      if(status !== 200)
        throw new Error(JSON.stringify(data,null,4));
      return data;
    })
    .catch(error=>{
      console.error(`Request Failed: ${url}`);
      throw error;
    });

const getLatestGitHubTag = async (gitHubRepo) =>
  get(`https://api.github.com/repos/${gitHubRepo}/tags`)
    .then(response=>response[0].name);

const getLatestNpmVersion = async (dependencyName)=>
  get(`https://registry.npmjs.org/${dependencyName}/`)
    .then(response=>response['dist-tags'].latest);

let currentRepo = undefined;

Promise.all(
  config.split('\n').map(async (line)=>{
    currentRepo = line.match(reExtractRepo)?.groups.gitHubRepo ?? currentRepo;
    let localCurrentRepo = currentRepo;

    // Updating Hook
    const currentRev = line.match(reRev)?.groups?.version;
    if(typeof currentRev !== 'undefined'){
      const latestTag = await getLatestGitHubTag(localCurrentRepo);
      if(currentRev === latestTag)
        return line;
      console.log(
        `Updating ${localCurrentRepo}' from ${currentRev} to ${latestTag}`
      );
      return line.replace(reRev, `$<base>${latestTag}`);
    }

    // Updating additional NPM dependency
    const dependency = line.match(reExtractDependency)?.groups;
    if(typeof dependency !== 'undefined'){
      const {dependencyName, currentVersion} = dependency;
      const latestVersion = await getLatestNpmVersion(dependencyName);
      if(currentVersion === `${latestVersion}`)
        return line;
      console.log(
        `Updating ${dependencyName}' from ${currentVersion} `
        + `to ${latestVersion}`
      );
      return line.replace(reReplaceVersion, `$<base>^${latestVersion}`);
    }
    return line;
  })
).then((results)=>
  fs.promises.writeFile(configFileLocation,results.join('\n'))
);
